feat(tianmu): support add integer/decimal column DDL with default value. (#1187)

lujiashun opened the pull request:
<!--

Thank you for contributing to StoneDB!

PR Title Format: <type>(<scope>): description to this pr (#issue_id)
e.g.
fix(util): fix sth..... (#1)

-->

## Summary about this PR
<!--

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
e.g.:
Issue: close #1

-->

Issue Number: close #1187

[summary]
check the default value of field,please see tianmu_attr.cpp for details:
1. if field is integer type but not real type;
2. if field is integer type and also real type;
3. if field is not integer type;
## Tests Check List
<!-- At least one of next options must be included. -->

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog
<!-- At least one of next options must be included. -->

- [x] New Feature
- [ ] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation
<!-- At least one of next options must be included. -->

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

wisehead commented on storage/tianmu/core/tianmu_attr.cpp:
```diff
@@ -87,19 +87,42 @@ void TianmuAttr::Create(const fs::path &dir, const AttributeTypeInfo &ati, uint8
+  size_t no_nulls{no_rows};
+  int64_t min{0};
+  int64_t max{0};
+  bool int_not_null{false};
```
why not use AttributeTypeInfo::NotNull() to check whether the field is NULL? I guess int_not_null means has_default_value?

lujiashun replied:
1.not null flag check is also a solution. in the scenario the alter table defined not null or null(same with tianmu), the
output is NULL. I just want to use GetValueFromField(it deals the default value logic) function to deal with the scenario, the std::variants store the value, it has the interface to decide whether it is null,if not null, what is the default value.
2 The current solution is also a proper way,not to consider the null flag (or it has been considered in GetValueFromField)
3 the std::variant in gcc8 has bugs, what we supported is gcc9.3, if we do not upgrade debian gcc to 9.3, another solution is to use boost:variant, but not recommended. It is better to upgrade the debian CI to gcc9.3 @wisehead

wisehead commented on storage/tianmu/handler/ha_tianmu.cpp:
```diff
@@ -112,7 +112,7 @@ static core::Value GetValueFromField(Field *f) {
       if (!common::IsTimeStampZero(my_time)) {
         my_bool myb;
-        my_time_t secs_utc = current_txn_->Thd()->variables.time_zone->TIME_to_gmt_sec(&my_time, &myb);
+        my_time_t secs_utc = f->table->in_use->variables.time_zone->TIME_to_gmt_sec(&my_time, &myb);
```
please add some comments here, I don't quite understand the code changes here.

lujiashun replied:
f->table->in_use is to get the THD pionter, current_txn_->Thd()  is also to get the THD pointer. I want to store default value from the function  GetValueFromField, but in some senario, the cureent_txn is not initialized and pointed to null. I find f->table->in_use is not nullptr, so i change the code to  f->table->in_use.

wisehead commented on storage/tianmu/handler/ha_tianmu.cpp:
```diff
@@ -1634,7 +1634,13 @@ bool ha_tianmu::inplace_alter_table(TABLE *altered_table, Alter_inplace_info *ha
       std::vector<Field *> v_old(table_share->field, table_share->field + table_share->fields);
-      std::vector<Field *> v_new(altered_table->s->field, altered_table->s->field + altered_table->s->fields);
+
+      Field_iterator_table it_field;
```
please also add some comments here, does original code work? or you change it because it's just ugly?

lujiashun replied:
Because i want to get the default value from the filed, but altered_table->s->field's table pointer is null, can not to get the default value.  alter_table->field's table pointer is correct, so i use it to store the fields;

RingsC commented on mysql-test/suite/tianmu/t/issue1187.test:
```diff
@@ -0,0 +1,40 @@
+
+create table t3(id int,name varchar(5)) ;
+insert into t3 values(1,'AAA'),(2,'BBB');
+alter table t3 add column age int not null default 66;
```
Pls, add some case for errors. such as using wrong default value.   
Taking the following as an instance.
The column type is `char` but using integer as a default value, the type and default value is mismatch. or the default will lead to type overflow.

RingsC replied:
F.I.Y: Just only change the default value.
``ALTER TABLE table_name ALTER COLUMN column_name SET DEFAULT 'literal';``

lujiashun replied:
1. ok, i will try to add some wrong default value; maybe it will be intercepted by sql layer, i will test it.(confirmed, it wll be intercepted by sql layer, better not to add this mtr);
2. "Just only change the default value" is a related sceanrio, i will add this MTR.

RingsC commented on storage/tianmu/handler/ha_tianmu.cpp:
```diff
@@ -1634,7 +1635,13 @@ bool ha_tianmu::inplace_alter_table(TABLE *altered_table, Alter_inplace_info *ha
+      Field_iterator_table it_field;
+      std::vector<Field *> v_new;
+      for (it_field.set_table(altered_table); !it_field.end_of_fields(); it_field.next()) {
+        v_new.push_back(it_field.field());
```
`emplace_back` if it has move ctor or as your wish.

lujiashun replied:
OK, i will fix it.

RingsC commented on storage/tianmu/handler/ha_tianmu.h:
```diff
@@ -23,6 +23,7 @@
 
 namespace Tianmu {
 namespace handler {
+extern Tianmu::core::Value GetValueFromField(Field *f);
```
Or perhaps, put it in .cpp, in which we will use this function, is a better way. And, will not contaminate the other files which includes this .h file.

lujiashun replied:
OK, i will fix it;

RingsC commented on storage/tianmu/core/value.h:
```diff
@@ -18,42 +18,87 @@
 
 namespace Tianmu {
 namespace core {
+// std::variant has some bug in gcc8
```
Pls, gives more detail info on this bug.  what behavior will happen on GCC 8.

lujiashun replied:
the bug is similar to this but not the same：https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85183 ，core::value has a member which type is  std::variant;  if AttributeTypeInfo has a memeber whose type is core:value, will report many compiler error that can not find matched (copy assignemend?  not the std::variant class ,its internal class), if upgrade to GCC9.3, it work well. GCC8 implementation for std::variant is not stable, So replace  std::variant instead of boost::variant; Keep std::variant for GCC9.3 anyway.

RingsC commented on storage/tianmu/core/value.h:
```diff
@@ -18,42 +18,87 @@
-  std::string_view &GetStringView() { return std::get<std::string_view>(var); }
-  const std::string_view &GetStringView() const { return std::get<std::string_view>(var); }
+  void SetStringView(char *s, uint n) { var_ = std::string_view(s, n); }
+  std::string_view &GetStringView() { return boost::get<std::string_view>(var_); }
```
and try do this 
```
class Value {
xxx
xxx
#if (__GCC__) && 

#else

#endif
xxxx
} ; 
```
is a better one. Now, this impl has a bunch of redundent code.

lujiashun replied:
OK, I will optimize this code to reduce redundant code.

RingsC commented on storage/tianmu/core/engine.cpp:
```diff
@@ -635,7 +635,16 @@ std::shared_ptr<TableOption> Engine::GetTableOption(const std::string &table, TA
+  bitmap_set_bit(field.table->write_set, field.field_index);
+  bitmap_set_bit(field.table->read_set, field.field_index);
+  field.set_default();
+  ati.SetDefaultValue(&field);
```
This line code was not covered by test.  but this line is the core code for this PR.

lujiashun replied:
The coverage code tool sometimes  does not take effect, actually the core code is runned,otherwise this PR does not work.

RingsC replied:
> The coverage code tool sometimes does not take effect, actually the core code is runned,otherwise this PR does not work.

I see.

Commit:
feat(tianmu): support add integer/decimal column DDL with default value. (#1187)

[summary]
check the default value of field,please see tianmu_attr.cpp for details:
1 if field is integer type but not real type;
2 if field is integer type and also real type;
2if field is not integer type;

DandreChen commented on mysql-test/suite/tianmu/t/issue1187.test:
```diff
@@ -0,0 +1,45 @@
+
+create table t2(id int,name varchar(5)) ;
+insert into t2 values(1,'AAA'),(2,'BBB');
+alter table t2 add column age int not null;
```
when the type is varchar not null，what will happen？

DandreChen replied:
if it is convenient，please fix it together when the type is varchar not null
as below case:
```
create table ttt(id int,name varchar(5));
insert into ttt values(1,'AAA'),(2,'BBB');
alter table ttt add column age varchar(5) not null;
select * from ttt;
```
